src/utils/transactions.ts

```diff
@@ -130,6 +132,40 @@ export const getMultiSendTxs = (
     .filter(Boolean) as MetaTransactionData[]
 }
 
+export const _isValidMultiSend = async (safe: SafeInfo, txs: MetaTransactionData[]): Promise<boolean> => {
+  const multiSendContract = getMultiSendCallOnlyContractInstance(safe.chainId, safe.version)
+  const data = encodeMultiSendData(txs)
+
+  try {
+    await multiSendContract.contract.callStatic.multiSend(data)
+    return true
+  } catch (err) {
+    return false
+  }
+}
+
+export const getValidBatch = async (
+  txs: TransactionDetails[],
+  safe: SafeInfo,
+  chain: ChainInfo,
+): Promise<MetaTransactionData[]> => {
+  const safeAddress = safe.address.value
+  const safeVersion = safe.version || LATEST_SAFE_VERSION
+
+  const multiSendTxs = getMultiSendTxs(txs, chain, safeAddress, safeVersion)
+
+  const isValidBatch = await _isValidMultiSend(safe, multiSendTxs)
+
+  if (isValidBatch) {
+    return multiSendTxs
+  }
+
+  const validity = await Promise.all(multiSendTxs.map((_, i) => _isValidMultiSend(safe, multiSendTxs.slice(0, i))))
```

Comment:
This will always test all possible tx combinations.
I think the most request number efficient implemention would be through binary search:

```ts
  let idx = multiSendTxs.length
  let bestBatch: MetaTransactionData[] = []
  while (idx > 0 && idx <= multiSendTxs.length) {
    const isValid = await _isValidMultiSend(safe, multiSendTxs.slice(0, idx))
    if (isValid) {
      bestBatch = multiSendTxs.slice(0, idx)
      idx += Math.floor(idx / 2)
    } else {
      idx = Math.floor(idx / 2)
    }
  }
  return bestBatch
```